File: modern-ui/electron/install-registry.ts
/**
 * Register this Zeus install with a URL you control (e.g. Supabase Edge Function; see /supabase/README.md).
 * POST { machineId, macAddress, version, os, arch }: auto-send at most once per 24h. "Send now" bypasses
 * that but is throttled to once per SEND_NOW_COOLDOWN_MS.
 */
import { app } from 'electron'
import fs from 'fs'
import os from 'os'
import path from 'path'
import { machineIdSync } from 'node-machine-id'

const STATE_FILE = 'install-registry.json'
const MIN_INTERVAL_MS = 24 * 60 * 60 * 1000
/** Minimum time between two "Send now" attempts (stops button spam; separate from 24h auto-send). */
const SEND_NOW_COOLDOWN_MS = 60_000

export type InstallRegistryPayload = {
  machineId: string
  macAddress: string | null
  version: string
  os: string
  arch: string
}

type RegistryState = {
  enabled: boolean
  lastSentAt: number | null
  /** Set when a forced ("Send now") request is started; throttles the next force. */
  lastManualSendAt: number | null
  lastSentStatus: 'success' | 'error' | 'disabled' | 'skipped' | null
  lastSentError: string | null
}

function getRegistryUrl(): string {
  return (process.env['INSTALL_REGISTRY_URL'] || '').trim()
}

function getRegistryToken(): string {
  return (process.env['REGISTRY_TOKEN'] || '').trim()
}

function statePath(): string {
  return path.join(app.getPath('userData'), STATE_FILE)
}

function defaultState(): RegistryState {
  return {
    enabled: true,
    lastSentAt: null,
    lastManualSendAt: null,
    lastSentStatus: null,
    lastSentError: null,
  }
}

function readState(): RegistryState {
  const p = statePath()
  if (!fs.existsSync(p)) return defaultState()
  try {
    const raw = fs.readFileSync(p, 'utf-8')
    const parsed = JSON.parse(raw) as Partial<RegistryState>
    if (!parsed || typeof parsed !== 'object' || typeof parsed.enabled !== 'boolean') {
      return defaultState()
    }
    return {
      enabled: parsed.enabled,
      lastSentAt: typeof parsed.lastSentAt === 'number' ? parsed.lastSentAt : null,
      lastManualSendAt: typeof parsed.lastManualSendAt === 'number' ? parsed.lastManualSendAt : null,
      lastSentStatus: parsed.lastSentStatus ?? null,
      lastSentError: typeof parsed.lastSentError === 'string' ? parsed.lastSentError : null,
    }
  } catch {
    return defaultState()
  }
}

function writeState(s: RegistryState) {
  try {
    fs.writeFileSync(statePath(), JSON.stringify(s, null, 2), 'utf-8')
  } catch (err) {
    console.warn('[install-registry] failed to persist state:', err)
  }
}

function patchState(partial: Partial<RegistryState>): RegistryState {
  const next = { ...readState(), ...partial }
  writeState(next)
  return next
}

/** First non-loopback interface with a usable MAC. */
export function getPrimaryMacAddress(): string | null {
  const ifaces = os.networkInterfaces()
  for (const name of Object.keys(ifaces).sort()) {
    const addrs = ifaces[name]
    if (!addrs) continue
    for (const a of addrs) {
      if (a.internal) continue
      const m = a.mac
      if (m && m !== '00:00:00:00:00:00') return m
    }
  }
  return null
}

export function getMachineIdString(): string {
  try {
    return machineIdSync()
  } catch (err) {
    console.warn('[install-registry] machineIdSync failed:', err)
    return ''
  }
}

function currentPayload(): InstallRegistryPayload {
  return {
    machineId: getMachineIdString(),
    macAddress: getPrimaryMacAddress(),
    version: app.getVersion(),
    os: process.platform,
    arch: process.arch,
  }
}

export type InstallRegistryStatus = {
  enabled: boolean
  endpoint: string | null
  lastSentAt: number | null
  /** Remaining ms before "Send now" is allowed; 0 if ready. */
  sendNowAfterMs: number
  lastSentStatus: 'success' | 'error' | 'disabled' | 'skipped' | null
  lastSentError: string | null
  hasToken: boolean
  nextPayload: InstallRegistryPayload
}

function sendNowCooldownRemainingMs(s: RegistryState): number {
  if (s.lastManualSendAt == null) return 0
  return Math.max(0, SEND_NOW_COOLDOWN_MS - (Date.now() - s.lastManualSendAt))
}

export function getInstallRegistryStatus(): InstallRegistryStatus {
  const url = getRegistryUrl()
  const s = readState()
  return {
    enabled: s.enabled,
    endpoint: url || null,
    lastSentAt: s.lastSentAt,
    sendNowAfterMs: sendNowCooldownRemainingMs(s),
    lastSentStatus: s.lastSentStatus,
    lastSentError: s.lastSentError,
    hasToken: Boolean(getRegistryToken()),
    nextPayload: currentPayload(),
  }
}

export function setInstallRegistryEnabled(value: boolean): boolean {
  patchState({ enabled: value })
  return readState().enabled
}

export function getInstallRegistryEnabled(): boolean {
  return readState().enabled
}

export async function sendInstallRegistry(
  opts: { force?: boolean } = {},
): Promise<{
  status: 'success' | 'error' | 'disabled' | 'skipped'
  error?: string
  sendNowAfterMs?: number
  payload?: InstallRegistryPayload
}> {
  const url = getRegistryUrl()
  if (!url) {
    patchState({ lastSentStatus: 'disabled', lastSentError: 'INSTALL_REGISTRY_URL not set' })
    return { status: 'disabled', error: 'INSTALL_REGISTRY_URL not set' }
  }
  if (!getInstallRegistryEnabled()) {
    patchState({ lastSentStatus: 'disabled', lastSentError: 'disabled by user' })
    return { status: 'disabled', error: 'disabled by user' }
  }

  const payload = currentPayload()
  if (!payload.machineId) {
    patchState({ lastSentStatus: 'error', lastSentError: 'no machine id' })
    return { status: 'error', error: 'no machine id', payload }
  }

  const now = Date.now()
  const s = readState()
  if (opts.force) {
    if (s.lastManualSendAt != null && now - s.lastManualSendAt < SEND_NOW_COOLDOWN_MS) {
      const sendNowAfterMs = Math.max(0, SEND_NOW_COOLDOWN_MS - (now - s.lastManualSendAt))
      return {
        status: 'skipped',
        error: 'Send now: please wait before sending again.',
        sendNowAfterMs,
        payload,
      }
    }
    patchState({ lastManualSendAt: now })
  }
  if (!opts.force && s.lastSentAt && now - s.lastSentAt < MIN_INTERVAL_MS) {
    return { status: 'skipped', payload }
  }

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  }
  const token = getRegistryToken()
  if (token) {
    headers['Authorization'] = `Bearer ${token}`
  }

  try {
    const res = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(payload),
    })
    if (!res.ok) {
      const errText = await res.text().catch(() => res.statusText)
      const msg = `HTTP ${res.status}: ${errText.slice(0, 200)}`
      patchState({ lastSentAt: now, lastSentStatus: 'error', lastSentError: msg })
      return { status: 'error', error: msg, payload }
    }
    patchState({ lastSentAt: now, lastSentStatus: 'success', lastSentError: null })
    return { status: 'success', payload }
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    patchState({ lastSentAt: now, lastSentStatus: 'error', lastSentError: msg })
    return { status: 'error', error: msg, payload }
  }
}

File: modern-ui/src-tauri/src/backend/install_registry.rs
use super::paths;
use reqwest::Client;
use serde_json::{json, Value};
use std::fs;
use std::time::{SystemTime, UNIX_EPOCH};

const STATE_FILE: &str = "install-registry.json";
const MIN_INTERVAL_MS: u64 = 24 * 60 * 60 * 1000;
const SEND_NOW_COOLDOWN_MS: u64 = 60_000;

#[derive(serde::Serialize, serde::Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
struct InstallRegistryPayload {
    machine_id: String,
    mac_address: Option<String>,
    version: String,
    os: String,
    arch: String,
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
struct RegistryState {
    #[serde(default = "default_enabled")]
    enabled: bool,
    last_sent_at: Option<u64>,
    last_manual_send_at: Option<u64>,
    last_sent_status: Option<String>,
    last_sent_error: Option<String>,
}

fn default_enabled() -> bool {
    true
}

fn registry_url() -> String {
    std::env::var("INSTALL_REGISTRY_URL")
        .unwrap_or_else(|_| option_env!("ZEUS_EMBED_INSTALL_REGISTRY_URL").unwrap_or("").to_string())
        .trim()
        .to_string()
}

fn registry_token() -> String {
    std::env::var("REGISTRY_TOKEN")
        .unwrap_or_else(|_| option_env!("ZEUS_EMBED_REGISTRY_TOKEN").unwrap_or("").to_string())
        .trim()
        .to_string()
}

fn state_path() -> std::io::Result<std::path::PathBuf> {
    Ok(paths::ensure_data_dir()?.join(STATE_FILE))
}

fn read_state() -> RegistryState {
    let path = match state_path() {
        Ok(p) => p,
        Err(_) => return RegistryState {
            enabled: true,
            last_sent_at: None,
            last_manual_send_at: None,
            last_sent_status: None,
            last_sent_error: None,
        },
    };
    if !path.exists() {
        return RegistryState {
            enabled: true,
            last_sent_at: None,
            last_manual_send_at: None,
            last_sent_status: None,
            last_sent_error: None,
        };
    }
    fs::read_to_string(path)
        .ok()
        .and_then(|raw| serde_json::from_str(&raw).ok())
        .unwrap_or(RegistryState {
            enabled: true,
            last_sent_at: None,
            last_manual_send_at: None,
            last_sent_status: None,
            last_sent_error: None,
        })
}

fn write_state(s: &RegistryState) {
    if let Ok(path) = state_path() {
        let _ = fs::write(path, serde_json::to_string_pretty(s).unwrap_or_default());
    }
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

fn primary_mac() -> Option<String> {
    mac_address::get_mac_address()
        .ok()
        .flatten()
        .map(|m| m.to_string())
}

fn machine_id_string() -> String {
    machine_uid::get().unwrap_or_default()
}

fn current_payload() -> InstallRegistryPayload {
    InstallRegistryPayload {
        machine_id: machine_id_string(),
        mac_address: primary_mac(),
        version: env!("CARGO_PKG_VERSION").to_string(),
        os: std::env::consts::OS.to_string(),
        arch: std::env::consts::ARCH.to_string(),
    }
}

fn send_now_cooldown_ms(s: &RegistryState) -> u64 {
    match s.last_manual_send_at {
        None => 0,
        Some(t) => SEND_NOW_COOLDOWN_MS.saturating_sub(now_ms().saturating_sub(t)),
    }
}

pub fn get_status() -> Value {
    let url = registry_url();
    let s = read_state();
    json!({
        "enabled": s.enabled,
        "endpoint": if url.is_empty() { Value::Null } else { Value::String(url) },
        "lastSentAt": s.last_sent_at,
        "sendNowAfterMs": send_now_cooldown_ms(&s),
        "lastSentStatus": s.last_sent_status,
        "lastSentError": s.last_sent_error,
        "hasToken": !registry_token().is_empty(),
        "nextPayload": current_payload(),
    })
}

pub fn set_enabled(value: bool) -> Value {
    let mut s = read_state();
    s.enabled = value;
    write_state(&s);
    Value::Bool(s.enabled)
}

pub async fn send_now(client: &Client, force: bool) -> Value {
    let url = registry_url();
    if url.is_empty() {
        let mut s = read_state();
        s.last_sent_status = Some("disabled".into());
        s.last_sent_error = Some("INSTALL_REGISTRY_URL not set".into());
        write_state(&s);
        return json!({ "status": "disabled", "error": "INSTALL_REGISTRY_URL not set" });
    }
    let mut s = read_state();
    if !s.enabled {
        s.last_sent_status = Some("disabled".into());
        s.last_sent_error = Some("disabled by user".into());
        write_state(&s);
        return json!({ "status": "disabled", "error": "disabled by user" });
    }

    let payload = current_payload();
    if payload.machine_id.is_empty() {
        s.last_sent_status = Some("error".into());
        s.last_sent_error = Some("no machine id".into());
        write_state(&s);
        return json!({ "status": "error", "error": "no machine id", "payload": payload });
    }

    let now = now_ms();
    if force {
        if let Some(last) = s.last_manual_send_at {
            if now.saturating_sub(last) < SEND_NOW_COOLDOWN_MS {
                return json!({
                    "status": "skipped",
                    "error": "Send now: please wait before sending again.",
                    "sendNowAfterMs": send_now_cooldown_ms(&s),
                    "payload": payload,
                });
            }
        }
        s.last_manual_send_at = Some(now);
        write_state(&s);
    } else if let Some(last) = s.last_sent_at {
        if now.saturating_sub(last) < MIN_INTERVAL_MS {
            return json!({ "status": "skipped", "payload": payload });
        }
    }

    let mut req = client.post(&url).header("Content-Type", "application/json");
    let token = registry_token();
    if !token.is_empty() {
        req = req.header("Authorization", format!("Bearer {token}"));
    }

    match req.json(&payload).send().await {
        Ok(res) => {
            let status = res.status();
            if status.is_success() {
                let mut s = read_state();
                s.last_sent_at = Some(now);
                s.last_sent_status = Some("success".into());
                s.last_sent_error = None;
                write_state(&s);
                json!({ "status": "success", "payload": payload })
            } else {
                let err_text = res.text().await.unwrap_or_default();
                let msg = format!(
                    "HTTP {}: {}",
                    status,
                    err_text.chars().take(200).collect::<String>()
                );
                let mut s = read_state();
                s.last_sent_at = Some(now);
                s.last_sent_status = Some("error".into());
                s.last_sent_error = Some(msg.clone());
                write_state(&s);
                json!({ "status": "error", "error": msg, "payload": payload })
            }
        }
        Err(e) => {
            let msg = e.to_string();
            let mut s = read_state();
            s.last_sent_at = Some(now);
            s.last_sent_status = Some("error".into());
            s.last_sent_error = Some(msg.clone());
            write_state(&s);
            json!({ "status": "error", "error": msg, "payload": payload })
        }
    }
}
